[PyTorchEdge] promote prim ops by using ops table for mobile runtime

pavithranrao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64816

Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

Commit:
[PyTorchEdge] promote prim ops by using ops table for mobile runtime



Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/mobile/function.h:
```diff
@@ -2,6 +2,7 @@
 
 #include <ATen/core/function_schema.h>
 #include <ATen/core/ivalue.h>
+#include <torch/csrc/jit/mobile/prim_ops_registery.h>
```
If it's only used in function.cpp, it's better to move it to the cpp file.

iseeyuan commented on torch/csrc/jit/mobile/prim_ops_registery.h:
```diff
@@ -0,0 +1,35 @@
+std::function<void(Stack&)>& getPrimOpsFn(const std::string& name);
+
+class prim_op_fn_register {
+  std::string prim_ops_name_;
```
Could you try to remove this quickly and see if it still works?

iseeyuan commented on torch/csrc/jit/mobile/promoted_prim_ops.cpp:
```diff
@@ -0,0 +1,190 @@
+    mobile::prim_op_fn_register("prim::Uninitialized", unInitialized),
+    mobile::prim_op_fn_register("aten::to.prim_dtype", toPrimDType),
+    mobile::prim_op_fn_register("prim::is_cuda", isCuda)
+    // TODO: (@pavithran) size is overloaded with int[] and Tensor
```
It may be some legacy code. There are two operators: `aten::size` and `aten::Size`. The only difference is the cap `S`. You may need to look at which one need to be promoted.

iseeyuan reviewed (approved):
LGTM. Some minor inline comments.

Commit:
Update on "[PyTorchEdge] promote prim ops by using ops table for mobile runtime"


Differential Revision: [D30819926](https://our.internmc.facebook.com/intern/diff/D30819926/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30819926/)!

[ghstack-poisoned]
